ompi/mca/coll/base/coll_base_allreduce.c

```diff
@@ -1288,24 +1287,19 @@ int ompi_coll_base_allreduce_intra_allgather_reduce(const void *sbuf, void *rbuf
         return OMPI_ERR_OUT_OF_RESOURCE;
     }
 
-    if (commutative) {
-        ompi_datatype_copy_content_same_ddt(dtype, count, (char *) rbuf, (char *) send_buf);
-    }
-
     tmp_buf = tmp_buf_raw - gap;
 
     /* Requests for send to AND receive from everyone else */
     int reqs_needed = (comm_size - 1) * 2;
     reqs = ompi_coll_base_comm_get_reqs(module->base_data, reqs_needed);
 
-    ptrdiff_t incr = extent * count;
-    tmp_recv = (char *) tmp_buf;
+    const ptrdiff_t incr = extent * count;
 
-    /* Exchange data with peer processes */
+    /* Exchange data with peer processes, excluding self */
     int req_index = 0, peer_rank = 0;
     for (int i = 1; i < comm_size; ++i) {
-        peer_rank = (rank + i) % comm_size;
-        tmp_recv = tmp_buf + (peer_rank * incr);
+        peer_rank = (rank + i) % comm_size;      /* Start at the next rank */
```

Comment:
you can receive the rank 0 data directly in rbuf if you are not IN_PLACE. As in the code I provided.

Reply:
Thank you! Addressed in update.